src/fparser/two/tests/fortran2003/test_primary_r701.py

```diff
@@ -0,0 +1,196 @@
+
+'''
+import sys
+
+import pytest
+
+import fparser.two.Fortran2003 as f2003
+import fparser.two.utils
+
+
+def assert_subclass_parse(source, base_type, actual_type=None,
+                          expected_str=None):
+    '''Assert that the given source matches the given ``base_type``
+    and optionally the specific type that it should produce.
+    '''
+    obj = f2003.Primary(source)
+    # Note: Check that the type exists in the possible types, rather than
+    # checking the type is an instance of one of the possible types (and
+    # all of its ancestor types). This is a much more targetted test as a
+    # result.
+    assert type(obj) in possible_subclasses(base_type)
+
+    if actual_type:
+        assert isinstance(obj, actual_type)
+    else:
+        assert isinstance(obj, base_type)
+
+    if expected_str:
+        assert expected_str == str(obj)
+
+
+def possible_subclasses(node_type, _seen=None):
+    '''Given a type (e.g. Fortran2003.Primary), return all of the
+    subtypes that could have been matched after parsing.
+    '''
+    seen = _seen or []
+    subclasses = getattr(node_type, 'subclass_names', [])
+    if node_type not in seen:
+        seen.append(node_type)
+
+    for subclass_name in subclasses:
+        module = sys.modules[node_type.__module__]
+        subclass = getattr(module, subclass_name, None)
+        if subclass is not None and subclass not in seen:
+            seen.append(subclass)
+            possible_subclasses(subclass, seen)
+    return seen
+
+
+def test_Constant(f2003_create):
+    '''Test that Constant types are matched by Primary.
+    '''
+    assert_subclass_parse(
+        '1.2e-03', f2003.Constant,
+        actual_type=f2003.Real_Literal_Constant,
+        expected_str='1.2E-03')
+
+
+def test_Designator(f2003_create):
+    '''Test that Designator types are matched by Primary.
+    '''
+    assert_subclass_parse(
+        'array(1:5)', f2003.Designator,
+        actual_type=f2003.Array_Section,
+        expected_str='array(1 : 5)')
+
+
+def test_Array_Constructor(f2003_create):
+    '''Test that Array Constructor types are matched by Primary.
+    '''
+    assert_subclass_parse(
+        '[ 1.2, 2.3e + 2,    -5.1 e-3 ]', f2003.Array_Constructor,
+        actual_type=f2003.Array_Constructor,
+        expected_str='[1.2, 2.3E+2, - 5.1E-3]')
+
+
+def test_Structure_Constructor(f2003_create):
+    '''Test that Structure Constructor types are matched by Primary.
+    '''
+    # Note: The actual returned type is Part_Ref. With more context of what
+    # has already been parsed it is possible that this could change in the
+    # future. For instance, in this example "PERSON" could actually be the
+    # name of a function.
+    assert_subclass_parse(
+        'PERSON ( 12,   "Jones" )', f2003.Structure_Constructor,
+        actual_type=f2003.Part_Ref,
+        expected_str='PERSON(12, "Jones")')
+
+
+@pytest.mark.xfail(reason="Unable to reach Function_Reference when parsing")
```

Comment:
We have a recent (unwritten!) rule that all xfails should reference an issue (otherwise they might never get fixed). Could you please raise an issue and then point to it from within the text.

Reply:
Thanks (and for the other xfail tests too).